src/Utils/BeanDescriptor.php

```diff
@@ -373,38 +397,68 @@ private function generateBeanConstructor() : MethodGenerator
+            /** @var DirectForeignKeyMethodDescriptor $desc */
+            $desc = $this->checkForDuplicate(new DirectForeignKeyMethodDescriptor($fk, $this->table, $this->namingStrategy, $this->annotationParser, $this->beanNamespace));
+            $descriptors[] = $desc;
         }
 
-        return $descriptors;
+        $this->directForeignKeysDescriptors = $descriptors;
+        return $this->directForeignKeysDescriptors;
     }
 
     /**
      * @return PivotTableMethodsDescriptor[]
      */
     private function getPivotTableDescriptors(): array
     {
+        if ($this->pivotTableDescriptors !== null) {
+            return $this->pivotTableDescriptors;
+        }
         $descs = [];
         foreach ($this->schemaAnalyzer->detectJunctionTables(true) as $table) {
             // There are exactly 2 FKs since this is a pivot table.
             $fks = array_values($table->getForeignKeys());
 
             if ($fks[0]->getForeignTableName() === $this->table->getName()) {
                 list($localFk, $remoteFk) = $fks;
-                $descs[] = new PivotTableMethodsDescriptor($table, $localFk, $remoteFk, $this->namingStrategy, $this->beanNamespace, $this->annotationParser);
+                /** @var PivotTableMethodsDescriptor $desc */
+                $desc = $this->checkForDuplicate(new PivotTableMethodsDescriptor($table, $localFk, $remoteFk, $this->namingStrategy, $this->beanNamespace, $this->annotationParser));
+                $descs[] = $desc;
             }
             if ($fks[1]->getForeignTableName() === $this->table->getName()) {
                 list($remoteFk, $localFk) = $fks;
-                $descs[] = new PivotTableMethodsDescriptor($table, $localFk, $remoteFk, $this->namingStrategy, $this->beanNamespace, $this->annotationParser);
+                /** @var PivotTableMethodsDescriptor $desc */
+                $desc = $this->checkForDuplicate(new PivotTableMethodsDescriptor($table, $localFk, $remoteFk, $this->namingStrategy, $this->beanNamespace, $this->annotationParser));
+                $descs[] = $desc;
             }
         }
 
-        return $descs;
+        $this->pivotTableDescriptors = $descs;
+        return $this->pivotTableDescriptors;
+    }
+
+    /**
+     * Check the method name isn't already used and flag the $descriptor to use its alternative name if it is the case
+     */
+    public function checkForDuplicate(MethodDescriptorInterface $descriptor): MethodDescriptorInterface
+    {
+        $name = $descriptor->getName();
+        if (!isset($this->descriptorsByMethodName[$name])) {
+            $this->descriptorsByMethodName[$name] = 0;
+        }
+        $this->descriptorsByMethodName[$name] ++;
+        if ($this->descriptorsByMethodName[$name] > 1) {
+            $descriptor->useAlternativeName();
+        }
+        return $descriptor;
```

Comment:
This might introduce a regression in generated code.

Have a careful look at the previous implementation:

If 2 descriptors are in conflict, then the 2 descriptors are using the alternative name.
In your implementation, only the second descriptor is using the alternative name. The first descriptor is untouched.

This is a problem because:

- it is a potential BC break.
- having the 2 descriptors renamed is important. Indeed. imagine some code that has been working for quite some time. Suddenly, the developer adds a table that creates a conflict. If the 2 methods are renamed, we can detect the error really quickly with PHPStan (one method we were calling does not exist anymore). On the other end, if only one method is renamed, we might be renaming the function we were previously using and replacing that function with another one. This is a problem that might be way harder to notice.

In the end, I think `$this->descriptorsByMethodName` should contain the list of all descriptors for this method name (`@var array<string, MethodDescriptorInterface[]>`. That way, you can call `useAlternativeName` on all descriptors in conflict, not just on the last one.